File: tests/dataset/job/compiler/c/q3.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
  int len;
  int *data;
} list_int;
static list_int list_int_create(int len) {
  list_int l;
  l.len = len;
  l.data = calloc(len, sizeof(int));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}
typedef struct {
  int len;
  double *data;
} list_float;
static list_float list_float_create(int len) {
  list_float l;
  l.len = len;
  l.data = calloc(len, sizeof(double));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}
typedef struct {
  int len;
  char **data;
} list_string;
static list_string list_string_create(int len) {
  list_string l;
  l.len = len;
  l.data = calloc(len, sizeof(char *));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}
typedef struct {
  int len;
  list_int *data;
} list_list_int;
static list_list_int list_list_int_create(int len) {
  list_list_int l;
  l.len = len;
  l.data = calloc(len, sizeof(list_int));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}
static char *_min_string(list_string v) {
  if (v.len == 0)
    return "";
  char *m = v.data[0];
  for (int i = 1; i < v.len; i++)
    if (strcmp(v.data[i], m) < 0)
      m = v.data[i];
  return m;
}
static int contains_list_string(list_string v, char *item) {
  for (int i = 0; i < v.len; i++)
    if (strcmp(v.data[i], item) == 0)
      return 1;
  return 0;
}
static int contains_string(char *s, char *sub) {
  return strstr(s, sub) != NULL;
}
static void _json_int(int v) { printf("%d", v); }
static void _json_float(double v) { printf("%g", v); }
static void _json_string(char *s) { printf("\"%s\"", s); }
static void _json_list_int(list_int v) {
  printf("[");
  for (int i = 0; i < v.len; i++) {
    if (i > 0)
      printf(",");
    _json_int(v.data[i]);
  }
  printf("]");
}
static void _json_list_float(list_float v) {
  printf("[");
  for (int i = 0; i < v.len; i++) {
    if (i > 0)
      printf(",");
    _json_float(v.data[i]);
  }
  printf("]");
}
static void _json_list_string(list_string v) {
  printf("[");
  for (int i = 0; i < v.len; i++) {
    if (i > 0)
      printf(",");
    _json_string(v.data[i]);
  }
  printf("]");
}
static void _json_list_list_int(list_list_int v) {
  printf("[");
  for (int i = 0; i < v.len; i++) {
    if (i > 0)
      printf(",");
    _json_list_int(v.data[i]);
  }
  printf("]");
}
typedef struct {
  char *movie_title;
} TmpItem;
typedef struct {
  int len;
  TmpItem *data;
} list_TmpItem;
static list_TmpItem list_TmpItem_create(int len) {
  list_TmpItem l;
  l.len = len;
  l.data = calloc(len, sizeof(TmpItem));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}

typedef struct {
  int id;
  char *keyword;
} KeywordItem;
typedef struct {
  int len;
  KeywordItem *data;
} list_KeywordItem;
static list_KeywordItem list_KeywordItem_create(int len) {
  list_KeywordItem l;
  l.len = len;
  l.data = calloc(len, sizeof(KeywordItem));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}

typedef struct {
  int movie_id;
  char *info;
} Movie_infoItem;
typedef struct {
  int len;
  Movie_infoItem *data;
} list_Movie_infoItem;
static list_Movie_infoItem list_Movie_infoItem_create(int len) {
  list_Movie_infoItem l;
  l.len = len;
  l.data = calloc(len, sizeof(Movie_infoItem));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}

typedef struct {
  int movie_id;
  int keyword_id;
} Movie_keywordItem;
typedef struct {
  int len;
  Movie_keywordItem *data;
} list_Movie_keywordItem;
static list_Movie_keywordItem list_Movie_keywordItem_create(int len) {
  list_Movie_keywordItem l;
  l.len = len;
  l.data = calloc(len, sizeof(Movie_keywordItem));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}

typedef struct {
  int id;
  char *title;
  int production_year;
} TitleItem;
typedef struct {
  int len;
  TitleItem *data;
} list_TitleItem;
static list_TitleItem list_TitleItem_create(int len) {
  list_TitleItem l;
  l.len = len;
  l.data = calloc(len, sizeof(TitleItem));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}

typedef struct {
  char *movie_title;
} ResultItem;
typedef struct {
  int len;
  ResultItem *data;
} list_ResultItem;
static list_ResultItem list_ResultItem_create(int len) {
  list_ResultItem l;
  l.len = len;
  l.data = calloc(len, sizeof(ResultItem));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}

static list_int test_Q3_returns_lexicographically_smallest_sequel_title_result;
static void test_Q3_returns_lexicographically_smallest_sequel_title() {
  list_int tmp1 = list_int_create(1);
  tmp1.data[0] = (TmpItem){.movie_title = "Alpha"};
  int tmp2 = 1;
  if (test_Q3_returns_lexicographically_smallest_sequel_title_result.len !=
      tmp1.len) {
    tmp2 = 0;
  } else {
    for (int i3 = 0;
         i3 <
         test_Q3_returns_lexicographically_smallest_sequel_title_result.len;
         i3++) {
      if (test_Q3_returns_lexicographically_smallest_sequel_title_result
              .data[i3] != tmp1.data[i3]) {
        tmp2 = 0;
        break;
      }
    }
  }
  if (!(tmp2)) {
    fprintf(stderr, "expect failed\n");
    exit(1);
  }
}

int main() {
  KeywordItem tmp4_data[] = {
      (KeywordItem){.id = 1, .keyword = "amazing sequel"},
      (KeywordItem){.id = 2, .keyword = "prequel"}};
  list_KeywordItem tmp4 = {2, tmp4_data};
  list_KeywordItem keyword = tmp4;
  Movie_infoItem tmp5_data[] = {
      (Movie_infoItem){.movie_id = 10, .info = "Germany"},
      (Movie_infoItem){.movie_id = 30, .info = "Sweden"},
      (Movie_infoItem){.movie_id = 20, .info = "France"}};
  list_Movie_infoItem tmp5 = {3, tmp5_data};
  list_Movie_infoItem movie_info = tmp5;
  Movie_keywordItem tmp6_data[] = {
      (Movie_keywordItem){.movie_id = 10, .keyword_id = 1},
      (Movie_keywordItem){.movie_id = 30, .keyword_id = 1},
      (Movie_keywordItem){.movie_id = 20, .keyword_id = 1},
      (Movie_keywordItem){.movie_id = 10, .keyword_id = 2}};
  list_Movie_keywordItem tmp6 = {4, tmp6_data};
  list_Movie_keywordItem movie_keyword = tmp6;
  TitleItem tmp7_data[] = {
      (TitleItem){.id = 10, .title = "Alpha", .production_year = 2006},
      (TitleItem){.id = 30, .title = "Beta", .production_year = 2008},
      (TitleItem){.id = 20, .title = "Gamma", .production_year = 2009}};
  list_TitleItem tmp7 = {3, tmp7_data};
  list_TitleItem title = tmp7;
  list_string tmp8 = list_string_create(8);
  tmp8.data[0] = "Sweden";
  tmp8.data[1] = "Norway";
  tmp8.data[2] = "Germany";
  tmp8.data[3] = "Denmark";
  tmp8.data[4] = "Swedish";
  tmp8.data[5] = "Denish";
  tmp8.data[6] = "Norwegian";
  tmp8.data[7] = "German";
  list_string allowed_infos = tmp8;
  list_string tmp9 = list_string_create(keyword.len * movie_keyword.len *
                                        movie_info.len * title.len);
  int tmp10 = 0;
  for (int tmp11 = 0; tmp11 < keyword.len; tmp11++) {
    KeywordItem k = keyword.data[tmp11];
    for (int tmp12 = 0; tmp12 < movie_keyword.len; tmp12++) {
      Movie_keywordItem mk = movie_keyword.data[tmp12];
      if (!(mk.keyword_id == k.id)) {
        continue;
      }
      for (int tmp13 = 0; tmp13 < movie_info.len; tmp13++) {
        Movie_infoItem mi = movie_info.data[tmp13];
        if (!(mi.movie_id == mk.movie_id)) {
          continue;
        }
        for (int tmp14 = 0; tmp14 < title.len; tmp14++) {
          TitleItem t = title.data[tmp14];
          if (!(t.id == mi.movie_id)) {
            continue;
          }
          if (!(contains_list_string(allowed_infos,
                                     contains_string(k.keyword, "sequel") &&
                                         mi.info) &&
                t.production_year > 2005 && mk.movie_id == mi.movie_id)) {
            continue;
          }
          tmp9.data[tmp10] = t.title;
          tmp10++;
        }
      }
    }
  }
  tmp9.len = tmp10;
  list_string candidate_titles = tmp9;
  ResultItem tmp15_data[] = {
      (ResultItem){.movie_title = _min_string(candidate_titles)}};
  list_ResultItem tmp15 = {1, tmp15_data};
  list_ResultItem result = tmp15;
  printf("[");
  for (int i16 = 0; i16 < result.len; i16++) {
    if (i16 > 0)
      printf(",");
    ResultItem it = result.data[i16];
    printf("{");
    _json_string("movie_title");
    printf(":");
    _json_string(it.movie_title);
    printf("}");
  }
  printf("]");
  test_Q3_returns_lexicographically_smallest_sequel_title_result = result;
  test_Q3_returns_lexicographically_smallest_sequel_title();
  return 0;
}

File: tests/dataset/job/compiler/go/q3.go
//go:build ignore

package main

import (
	"encoding/json"
	"fmt"
	"mochi/runtime/data"
	"reflect"
	"strings"
	"time"
)

func expect(cond bool) {
	if !cond {
		panic("expect failed")
	}
}

func formatDuration(d time.Duration) string {
	switch {
	case d < time.Microsecond:
		return fmt.Sprintf("%dns", d.Nanoseconds())
	case d < time.Millisecond:
		return fmt.Sprintf("%.1fµs", float64(d.Microseconds()))
	case d < time.Second:
		return fmt.Sprintf("%.1fms", float64(d.Milliseconds()))
	default:
		return fmt.Sprintf("%.2fs", d.Seconds())
	}
}

func printTestStart(name string) {
	fmt.Printf("   test %-30s ...", name)
}

func printTestPass(d time.Duration) {
	fmt.Printf(" ok (%s)\n", formatDuration(d))
}

func printTestFail(err error, d time.Duration) {
	fmt.Printf(" fail %v (%s)\n", err, formatDuration(d))
}

func test_Q3_returns_lexicographically_smallest_sequel_title() {
	expect(_equal(result, []map[string]string{map[string]string{"movie_title": "Alpha"}}))
}

type KeywordItem struct {
	Id      int    `json:"id"`
	Keyword string `json:"keyword"`
}

var keyword []KeywordItem

type Movie_infoItem struct {
	Movie_id int    `json:"movie_id"`
	Info     string `json:"info"`
}

var movie_info []Movie_infoItem

type Movie_keywordItem struct {
	Movie_id   int `json:"movie_id"`
	Keyword_id int `json:"keyword_id"`
}

var movie_keyword []Movie_keywordItem

type TitleItem struct {
	Id              int    `json:"id"`
	Title           string `json:"title"`
	Production_year int    `json:"production_year"`
}

var title []TitleItem
var allowed_infos []string
var candidate_titles []string

type ResultItem struct {
	Movie_title any `json:"movie_title"`
}

var result []ResultItem

func main() {
	failures := 0
	keyword = _cast[[]KeywordItem]([]KeywordItem{KeywordItem{
		Id:      1,
		Keyword: "amazing sequel",
	}, KeywordItem{
		Id:      2,
		Keyword: "prequel",
	}})
	movie_info = _cast[[]Movie_infoItem]([]Movie_infoItem{Movie_infoItem{
		Movie_id: 10,
		Info:     "Germany",
	}, Movie_infoItem{
		Movie_id: 30,
		Info:     "Sweden",
	}, Movie_infoItem{
		Movie_id: 20,
		Info:     "France",
	}})
	movie_keyword = _cast[[]Movie_keywordItem]([]Movie_keywordItem{
		Movie_keywordItem{
			Movie_id:   10,
			Keyword_id: 1,
		},
		Movie_keywordItem{
			Movie_id:   30,
			Keyword_id: 1,
		},
		Movie_keywordItem{
			Movie_id:   20,
			Keyword_id: 1,
		},
		Movie_keywordItem{
			Movie_id:   10,
			Keyword_id: 2,
		},
	})
	title = _cast[[]TitleItem]([]TitleItem{TitleItem{
		Id:              10,
		Title:           "Alpha",
		Production_year: 2006,
	}, TitleItem{
		Id:              30,
		Title:           "Beta",
		Production_year: 2008,
	}, TitleItem{
		Id:              20,
		Title:           "Gamma",
		Production_year: 2009,
	}})
	allowed_infos = []string{
		"Sweden",
		"Norway",
		"Germany",
		"Denmark",
		"Swedish",
		"Denish",
		"Norwegian",
		"German",
	}
	candidate_titles = func() []string {
		_res := []string{}
		for _, k := range keyword {
			for _, mk := range movie_keyword {
				if !(mk.Keyword_id == k.Id) {
					continue
				}
				for _, mi := range movie_info {
					if !(mi.Movie_id == mk.Movie_id) {
						continue
					}
					for _, t := range title {
						if !(t.Id == mi.Movie_id) {
							continue
						}
						if ((strings.Contains(k.Keyword, "sequel") && _contains[string](allowed_infos, mi.Info)) && (t.Production_year > 2005)) && (mk.Movie_id == mi.Movie_id) {
							if ((strings.Contains(k.Keyword, "sequel") && _contains[string](allowed_infos, mi.Info)) && (t.Production_year > 2005)) && (mk.Movie_id == mi.Movie_id) {
								_res = append(_res, t.Title)
							}
						}
					}
				}
			}
		}
		return _res
	}()
	result = _cast[[]ResultItem]([]ResultItem{ResultItem{Movie_title: _min(candidate_titles)}})
	func() { b, _ := json.Marshal(result); fmt.Println(string(b)) }()
	{
		printTestStart("Q3 returns lexicographically smallest sequel title")
		start := time.Now()
		var failed error
		func() {
			defer func() {
				if r := recover(); r != nil {
					failed = fmt.Errorf("%v", r)
				}
			}()
			test_Q3_returns_lexicographically_smallest_sequel_title()
		}()
		if failed != nil {
			failures++
			printTestFail(failed, time.Since(start))
		} else {
			printTestPass(time.Since(start))
		}
	}
	if failures > 0 {
		fmt.Printf("\n[FAIL] %d test(s) failed.\n", failures)
	}
}

func _cast[T any](v any) T {
	if tv, ok := v.(T); ok {
		return tv
	}
	var out T
	switch any(out).(type) {
	case int:
		switch vv := v.(type) {
		case int:
			return any(vv).(T)
		case float64:
			return any(int(vv)).(T)
		case float32:
			return any(int(vv)).(T)
		}
	case float64:
		switch vv := v.(type) {
		case int:
			return any(float64(vv)).(T)
		case float64:
			return any(vv).(T)
		case float32:
			return any(float64(vv)).(T)
		}
	case float32:
		switch vv := v.(type) {
		case int:
			return any(float32(vv)).(T)
		case float64:
			return any(float32(vv)).(T)
		case float32:
			return any(vv).(T)
		}
	}
	if m, ok := v.(map[any]any); ok {
		v = _convertMapAny(m)
	}
	data, err := json.Marshal(v)
	if err != nil {
		panic(err)
	}
	if err := json.Unmarshal(data, &out); err != nil {
		panic(err)
	}
	return out
}

func _contains[T comparable](s []T, v T) bool {
	for _, x := range s {
		if x == v {
			return true
		}
	}
	return false
}

func _convertMapAny(m map[any]any) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		key := fmt.Sprint(k)
		if sub, ok := v.(map[any]any); ok {
			out[key] = _convertMapAny(sub)
		} else {
			out[key] = v
		}
	}
	return out
}

func _equal(a, b any) bool {
	av := reflect.ValueOf(a)
	bv := reflect.ValueOf(b)
	if av.Kind() == reflect.Slice && bv.Kind() == reflect.Slice {
		if av.Len() != bv.Len() {
			return false
		}
		for i := 0; i < av.Len(); i++ {
			if !_equal(av.Index(i).Interface(), bv.Index(i).Interface()) {
				return false
			}
		}
		return true
	}
	if av.Kind() == reflect.Map && bv.Kind() == reflect.Map {
		if av.Len() != bv.Len() {
			return false
		}
		for _, k := range av.MapKeys() {
			bvVal := bv.MapIndex(k)
			if !bvVal.IsValid() {
				return false
			}
			if !_equal(av.MapIndex(k).Interface(), bvVal.Interface()) {
				return false
			}
		}
		return true
	}
	if (av.Kind() == reflect.Int || av.Kind() == reflect.Int64 || av.Kind() == reflect.Float64) &&
		(bv.Kind() == reflect.Int || bv.Kind() == reflect.Int64 || bv.Kind() == reflect.Float64) {
		return av.Convert(reflect.TypeOf(float64(0))).Float() == bv.Convert(reflect.TypeOf(float64(0))).Float()
	}
	return reflect.DeepEqual(a, b)
}

func _min(v any) any {
	if g, ok := v.(*data.Group); ok {
		v = g.Items
	}
	switch s := v.(type) {
	case []int:
		if len(s) == 0 {
			return 0
		}
		m := s[0]
		for _, n := range s[1:] {
			if n < m {
				m = n
			}
		}
		return m
	case []float64:
		if len(s) == 0 {
			return 0.0
		}
		m := s[0]
		for _, n := range s[1:] {
			if n < m {
				m = n
			}
		}
		return m
	case []string:
		if len(s) == 0 {
			return ""
		}
		m := s[0]
		for _, n := range s[1:] {
			if n < m {
				m = n
			}
		}
		return m
	case []any:
		if len(s) == 0 {
			return 0
		}
		switch s[0].(type) {
		case string:
			m := s[0].(string)
			for _, it := range s[1:] {
				v := it.(string)
				if v < m {
					m = v
				}
			}
			return m
		case int, int64, float64:
			var m float64
			var isFloat bool
			switch n := s[0].(type) {
			case int:
				m = float64(n)
			case int64:
				m = float64(n)
			case float64:
				m = n
				isFloat = true
			}
			for _, it := range s[1:] {
				switch v := it.(type) {
				case int:
					if float64(v) < m {
						m = float64(v)
					}
				case int64:
					if float64(v) < m {
						m = float64(v)
					}
				case float64:
					if v < m {
						m = v
					}
					isFloat = true
				}
			}
			if isFloat {
				return m
			}
			return int(m)
		default:
			panic("min() expects numbers or strings")
		}
	default:
		rv := reflect.ValueOf(v)
		if rv.Kind() == reflect.Slice {
			if rv.Len() == 0 {
				return 0
			}
			m := rv.Index(0).Interface()
			switch m.(type) {
			case int, int64, float64:
				items := make([]any, rv.Len())
				for i := 0; i < rv.Len(); i++ {
					items[i] = rv.Index(i).Interface()
				}
				return _min(items)
			}
		}
		panic("min() expects list or group")
	}
}
